src/crypto/tls/delegated_credentials.go

```diff
@@ -0,0 +1,533 @@
+		} else {
+			return nil, 0, fmt.Errorf("tls: unsupported delgation key type: %s", curveName)
+		}
+	case ed25519.PublicKey:
+		return pk, Ed25519, nil
+	default:
+		return nil, 0, fmt.Errorf("tls: unsupported delgation key type: %T", pk)
+	}
+}
+
+// marshal encodes the credential struct of the Delegated Credential.
+func (cred *Credential) marshal() ([]byte, error) {
+	ser := make([]byte, 4)
+	binary.BigEndian.PutUint32(ser, uint32(cred.validTime/time.Second))
+
+	var serAlgo [2]byte
+	binary.BigEndian.PutUint16(serAlgo[:], uint16(cred.expCertVerfAlgo))
+	ser = append(ser, serAlgo[:]...)
+
+	// Encode the public key
+	serPub, err := cred.marshalPublicKeyInfo()
+	if err != nil {
+		return nil, err
+	}
+	// Assert that the public key encoding is no longer than 2^24  bytes.
+	if len(serPub) > dcMaxPubLen {
+		return nil, errors.New("tls: public key length exceeds 2^24-1 limit")
+	}
+
+	var b cryptobyte.Builder
+	b.AddUint24(uint32(len(serPub)))
+	serLen := b.BytesOrPanic()
+	ser = append(ser, serLen[:]...)
+	ser = append(ser, serPub...)
+
+	return ser, nil
+}
+
+// unmarshalCredential decodes serialized bytes and returns a credential, if possible.
+func unmarshalCredential(ser []byte) (*Credential, error) {
+	if len(ser) < 9 {
+		return nil, errors.New("tls: delegated credential is not valid")
+	}
+
+	validTime := time.Duration(binary.BigEndian.Uint32(ser)) * time.Second
+	pubAlgo := SignatureScheme(binary.BigEndian.Uint16(ser[4:6]))
+
+	s := cryptobyte.String(ser[6:9])
+	var pubLen uint32
+	s.ReadUint24(&pubLen)
+
+	pubKey, err := x509.ParsePKIXPublicKey(ser[9:])
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ser[9:]) != int(pubLen) {
+		return nil, errors.New("tls: delegated credential is not valid")
+	}
+
+	return &Credential{validTime, pubAlgo, pubKey}, nil
+}
+
+// getCredentialLen returns the number of bytes comprising the serialized
+// credential struct inside the Delegated Credential.
+func getCredentialLen(ser []byte) (int, error) {
+	if len(ser) < 9 {
+		return 0, errors.New("tls: delegated credential is not valid")
+	}
+
+	// The validity time.
+	ser = ser[4:]
+	// The expCertVerfAlgo.
+	ser = ser[2:]
+
+	// The length of the Public Key.
+	s := cryptobyte.String(ser[:3])
+	var pubLen uint32
+	s.ReadUint24(&pubLen)
+	ser = ser[3:]
+	if len(ser) < int(pubLen) {
+		return 0, errors.New("tls: delegated credential is not valid")
+	}
+
+	return 9 + int(pubLen), nil
+}
+
+// getHash maps the SignatureScheme to its corresponding hash function.
+func getHash(scheme SignatureScheme) crypto.Hash {
+	switch scheme {
+	case ECDSAWithP256AndSHA256:
+		return crypto.SHA256
+	case ECDSAWithP384AndSHA384:
+		return crypto.SHA384
+	case ECDSAWithP521AndSHA512:
+		return crypto.SHA512
+	case Ed25519:
+		return crypto.SHA512
+	default:
+		return 0 //Unknown hash function
+	}
+}
+
+// getCurve maps the SignatureScheme to its corresponding elliptic.Curve.
+func getCurve(scheme SignatureScheme) elliptic.Curve {
```

Comment:
Again, it's a bit annoying that this function doesn't exist already in the TLS code.

Reply:
Agreed. `typeAndHashFromSignatureScheme` is similar but not exactly what is needed.